### PR TITLE
Refactor create/login/protected user use cases and implement passport sessions

### LINKED ISSUE #8: Substitute passport-jwt in auth with passport-local cookie sessions.
Currently, the login and creation use-cases both return signed JWTs upon successful execution. The `PassportUserAuthHandler` implemention of the `UserAuthHander` base class implements this. Since we're opting to establish a global auth microservice session, cookies are just more appropriate for this case than a JWT. This ticket involves removing our dependence on `passport-jwt` and switching to cookie-based authentication instead using `passport-local`. 

We already use `passport-local` in the codebase but only to setup the local strategies for authentication using Postgres; we don't have sessions/cookies supported. A great resource I found that discusses the ins and outs of this can be found [here](https://levelup.gitconnected.com/everything-you-need-to-know-about-the-passport-local-passport-js-strategy-633bbab6195). Note that the code examples provided in this sample may not be sufficient; feel free to research more into other peoples' implementations and how we could best integrate cookie-based auth with `passport-local` given our codebase design.

Among deprecating/editing the `PassportUserAuthHandler` class (that uses JWTs), tests will be needed to be updated to account for the removal of a token response type from the login and creation use-cases. Moreover, `passport-jwt` package dependencies should be removed as well.

Upon completion, both these endpoints should be extensively tested locally using Postman, downloadable [here](https://www.postman.com/product/api-client/).

### LINKED ISSUE #12: Modify login and creation use-cases to return authorization code.
This ticket should be done following the blocking ticket linked below (lots of overlap in the codebases, and we want to avoid a million merge conflicts, not to mention be able to easily test the use-cases).

As described [here](https://www.notion.so/db57fd8c4c5544af9f2875b4f73f80b2?v=e6dbf2632c3249e7839275e90e1d067e) under step 5, upon successful account creation and login, the use-cases should redirect the request to the service provider's authorization handling endpoint with the generated authorization code endpoint as the query parameter.

More specifically, the login/create use-cases should be modified to:

1. Accept the query parameters found [here](https://www.notion.so/db57fd8c4c5544af9f2875b4f73f80b2?v=e6dbf2632c3249e7839275e90e1d067e&p=d5150e473ded45f9b993189c8d4f2ac8).

(note that in the case of the login endpoint, the body of the request includes the plaintext username and password from the microservice frontend)

2. Respond with an HTTP 302 redirect to the `redirect_url`, with the authorization code endpoint included as a query parameter if and only if the following checks pass:

   1.  The `client_id` supplied corresponds to an existing entity in the `AuthSecret` database collection. Otherwise return the 403 Forbidden HTTP status.
   2. The `scope` query parameter supplied is strictly 'open_id'. Otherwise return the 400 Bad Request HTTP status.

These endpoints should appropriately sanitize any new query/body parameters (refer to how they already achieve this) and be sufficiently unit tested for any new behaviour.

### PR DESCRIPTION
**Pull request by ShreyasPrasad**
This PR addresses closes #12 and closes #8.

All token auth code has been removed.

### CODE REVIEW

File: package-lock.json
Comment (xujustinj) [inline package-lock.json]: in the main waterpark project we actually decided to drop `package-lock.json` after realizing how much of a pain in the * it is to manage

File: server/src/modules/users/application/use-cases/authenticate-user/authenticate-user-use-case.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/authenticate-user/authenticate-user-use-case.ts]:
```suggestion
      if (userByEmailAndPassword.isErr()) {
        return Result.err(new AuthenticateUserErrors.AuthenticationFailedError(email.value, userByEmailAndPassword.error.message))
      }
```

File: server/src/modules/users/application/use-cases/create-user/create-user-dto.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/create-user/create-user-dto.ts:35]: I'll make a PR to change these to io-ts validations once the mess that is API types is finalized

File: server/src/modules/users/application/use-cases/create-user/__tests__/create-user-use-case.test.unit.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/create-user/__tests__/create-user-use-case.test.unit.ts]:
```suggestion
```

I'm pretty sure this isn't needed since `createUserDTO.body` is already a reference to the `createUserDTOBody` object

Likewise all the places below

File: server/src/modules/users/application/use-cases/create-user/create-user-controller.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/create-user/create-user-controller.ts]: this is what you're actually testing for, right?

```suggestion
    if (Object.keys(req.params).length === 0) {
```

File: server/src/modules/users/application/use-cases/create-user/create-user-controller.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/create-user/create-user-controller.ts:40]: you might need to re-run Prettier

File: server/src/modules/users/application/use-cases/create-user/create-user-use-case.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/create-user/create-user-use-case.ts:32]: So from what I understand from this code, we either return the user directly, or we trigger a redirect if some special query params are given. I think this switching logic makes more sense going on the controller level (and then we have multiple use cases like your comment mentions).

File: server/src/modules/users/application/use-cases/login-user/login-user-use-case.ts
Comment (xujustinj) [inline server/src/modules/users/application/use-cases/login-user/login-user-use-case.ts:72]: a few nits

1. the nested ifs make this a bit hard to read, I'd try to fully separate the err and ok cases in separate if blocks
2. from the way your types are set up, I think `params.scope` is never undefined unless `params` is undefined
3. you can also maybe combine both of the checks containing `params && params.scope`

anyhoo I'm guessing you have a later todo planned to split the two different functionalities into separate use-cases, so fixing the code here would be pointless

File: server/src/shared/app/base-controller.ts
Comment (xujustinj) [inline server/src/shared/app/base-controller.ts:18]: nice, I'm gonna yoink that for the Waterpark BaseController too 😄

File: server/src/shared/app/param-list.ts
Comment (xujustinj) [inline server/src/shared/app/param-list.ts]:
```suggestion
        return url + '?' + this.list.map(({ name, value }) => `${name}=${value}`).join('&')
```

File: server/src/setup/http/express/versioned-api-router.ts
Comment (xujustinj) [inline server/src/setup/http/express/versioned-api-router.ts:13]: A few days ago I fell down a rabbit hole of whether to end urls with a trailing `/` or not, and didn't end up coming to any conclusions in favour of either way (people seem to disagree on which way to do it, but they also disagree on what the main reasons are for preferring their way). Do you have any reasons for prefering a trailing slash here? This is actually a pretty important thing to set a Loo Labs standard for.

File: server/src/shared/infra/db/entities/auth-code.entity.ts
Comment (xujustinj) [inline server/src/shared/infra/db/entities/auth-code.entity.ts:12]: when the time comes to actually use these entities with a DB (e.g. integration tests), remember to create a migration

File: server/src/shared/auth/implementations/mock-user-auth-handler.ts
Comment (xujustinj) [inline server/src/shared/auth/implementations/mock-user-auth-handler.ts:18]: Ooh I guess I missed this in the previous PR, but the idea of triggering error cases in mocks with specific inputs is really cool!

In fact, the whole dependency injection format of these handlers is awesome

**Review by xujustinj — COMMENTED** [review]
I'd be lying if I said I understood what these changes actually do, but there's no red flags in the code itself so I'm down to approve so auth is unblocked.

One question first: how was this tested?

File: server/src/shared/app/param-list.ts
Comment (ShreyasPrasad) [reply server/src/shared/app/param-list.ts]: Much nicer, thanks

File: server/src/modules/users/application/use-cases/login-user/login-user-use-case.ts
Comment (ShreyasPrasad) [reply server/src/modules/users/application/use-cases/login-user/login-user-use-case.ts:72]: Yeh that's the plan, don't like coupling the param-less code with the param-full code, but I think it's worth introducing a little tech debt right now for the sake of getting the other parts of auth working. Cleanup will follow.

File: server/src/modules/users/application/use-cases/create-user/create-user-use-case.ts
Comment (ShreyasPrasad) [reply server/src/modules/users/application/use-cases/create-user/create-user-use-case.ts:32]: yep, I like separating at the controller level. Will add a ticket for a refactor that will hopefully be addressed after a few weeks.

**Comment by ShreyasPrasad** [comment]
@xujustinj The endpoints were all tested in Postman. I created a collection that queries all the endpoints (/login, /create) with erroneous inputs and the happy paths as well). Might be worth adding to notion.

File: server/src/setup/http/express/versioned-api-router.ts
Comment (ShreyasPrasad) [reply server/src/setup/http/express/versioned-api-router.ts:13]: No specific preference, willing to lean either way.

File: server/src/shared/auth/implementations/mock-user-auth-handler.ts
Comment (ShreyasPrasad) [reply server/src/shared/auth/implementations/mock-user-auth-handler.ts:18]: Yeh that was motivated by dependency injection, definitely a cleaner way to organize it but this will do for the timebeing.

File: server/src/setup/http/express/versioned-api-router.ts
Comment (xujustinj) [reply server/src/setup/http/express/versioned-api-router.ts:13]: Dang, that makes both of us with no preference. @jeffzh4ng any thoughts?

Comment (j4orz) [reply server/src/setup/http/express/versioned-api-router.ts:13]: What did you find in the rabbit hole @xujustinj? I feel like the trailing slash indicates some sort of concatenation, and if the path is just `/users` then we should omit the trailing slash.

Comment (xujustinj) [reply server/src/setup/http/express/versioned-api-router.ts:13]: ### Colm Divilly's [blog](https://blog.cdivilly.com/2019/02/28/uri-trailing-slashes)

This post is a tl;dr of [a longer article by the same person](https://cdivilly.wordpress.com/2014/03/11/why-trailing-slashes-on-uris-are-important/). tl;dr of the tl;dr: trailing slashes make relative linking possible. Counterintuitively, relative links are not actually handled by concatenation.

### StackOverflow [thread](https://softwareengineering.stackexchange.com/questions/186959/trailing-slash-in-restful-api)

* REST has no opinion on this
* (mumbling about *nix file structures, which is where the slash-delimited URI format comes from in the first place)
* trailing slash is better for relative linking
* reserve trailing slashes for "directories" (e.g. `/clubs/` is a directory of clubs, but `/clubs/123abc` has no trailing slash because it's a single resource)
* trailing slash indicates that a resource is _not_ a file, which helps SEO in the pathological case where the end of a url may be misinterpreted as a file extension

### Issue 118 in [pyeve/eve](https://github.com/pyeve/eve)

(not linking directly because that'll trigger a backlink to this discussion)

Arguments go both ways, but tl;dr: it's very commonplace to leave out the trailing slash, but it's technically wrong to do so.

### Vladimir Sidorenko's [article](https://gearheart.io/articles/restful-api-design-best-practices/)

> AVOID USING A FORWARD SLASH AT THE END OF THE URL
> A trailing forward slash (“/”) has no semantic meaning and can lead to confusion.

This seems to go against the relative linking argument, but I guess what they're getting at here is that relative linking has no use in a REST API.

### Guy Levin's [blog](https://blog.restcase.com/7-rules-for-rest-api-uri-design/)

> Rule #1: A trailing forward slash (/) should not be included in URIs
> This is one the most important rules to follow as the last character within a URI’s path, a forward slash (/) adds no semantic value and may cause confusion. REST API’s should not expect a trailing slash and should not include them in the links that they provide to clients.

He also points out that a lot of tools/frameworks treat trailing-slash and no-trailing-slash equivalently (by redirecting one to the other). This is why the question of trailing slashes is so invisible and barely thought about in the modern web.

Comment (j4orz) [reply server/src/setup/http/express/versioned-api-router.ts:13]: Thanks for the detailed right up. I do like the semantic reasoning of directories vs single resources on the SO thread.  But on the other hand you have other resources like pyeve/eve saying it's technically wrong to have trailing slashes. Feel free to make the final decision on this one @xujustinj @ShreyasPrasad

**Comment by j4orz** [comment]
@ShreyasPrasad Can you enter "Closes #12" and Closes "#8" in the PR description instead of the zenhub urls so Github will auto-close those issues once the PR is merged?

File: server/src/shared/infra/db/entities/auth-code.entity.ts
Comment (j4orz) [inline server/src/shared/infra/db/entities/auth-code.entity.ts:11]: Can we resolve the TODO on L18 now like @KTong821's PR in Waterpark? https://github.com/loolabs/waterpark/pull/159

File: server/src/shared/app/base-controller.ts
Comment (j4orz) [reply server/src/shared/app/base-controller.ts:18]: time for the loo labs framework to be pulled out ;)

File: server/src/shared/auth/user-auth-handler.ts
Comment (j4orz) [inline server/src/shared/auth/user-auth-handler.ts]: Why is this just an `Error` now instead of `AppError.UnexpectedError`?

**Review by j4orz — CHANGES_REQUESTED** [review]
For the future I think we should split up our PRs — it'll take some more work up-front to plan out the PRs and the order they need to be merged in but they will be more manageable to review. Tbh I don't really know what's happening anymore 🥴

File: server/src/shared/auth/user-auth-handler.ts
Comment (ShreyasPrasad) [reply server/src/shared/auth/user-auth-handler.ts]: Good call, swapped to the more specific Error subclass.

File: server/src/modules/users/application/use-cases/create-user/create-user-controller.ts
Comment (ShreyasPrasad) [reply server/src/modules/users/application/use-cases/create-user/create-user-controller.ts:40]: Don't like depending on 'user'. This will change once we refactor into sep use cases.

**Review by xujustinj — APPROVED** [review]
After deleting `package-lock.json` you have to add

```
package-lock.json
```

to the top-level gitignore to make sure it doesn't come back. Can defer to a later PR (since there's lots of other cleanup to do anyway)

Rest "lgtm" though I still barely know what's going on 😅. I would wait for @jeffzh4ng's approval.

**Review by j4orz — APPROVED** [review]

**Pull request merged**
